### PR TITLE
Throttling, Watch and Download Refactor and debouncing file watcher

### PR DESCRIPTION
**Pull request by tanema**
Okay this turned into a much bigger change then I expected. I had to refactor the throttler because I figured out that it was not able to properly close the done channel when it was finished the jobs it was given.

### Changed

- refactored Foreman to close the worker queue when the job queue has been closed. This enables a cascading close making commands like replace and upload complete cleanly
- refactored download to use simpler functionality
- refactored watch now to use similar patterns to upload and replace so that they should be way more understandable. There is less duplicate code and just less code in general.
- added debouncing to the file watcher to fix issue #200. The problem was that for most text editors saving large files, there would be multiple save events on the file and we would only process the first event on that file. This lead to half saved files. Now I debounced and we are processing the last write event.
- added a `NewForman()` method to the theme client as well so that we are always creating Forman objects with the configured bucket sizes.
- added logging and central event log to the theme client since this was a huge pain cascading all of the events around different channels just to log them.
- took out an incredible amount of complexity that hopefully it will make the repo a lot more nice to work with.

### Keyboard Debouncing Benifits

The debouncing of file save events also may in the future might fix the following issues as well. This is because the errors were most likely from poorly uploaded files. So it woulld throw errors on syntax, or would remove the file because some editors switch files (~filename -> filename) and it would only get the remove event.
- fixes https://github.com/Shopify/themekit/issues/200
- fixes https://github.com/Shopify/themekit/issues/168
- fixes https://github.com/Shopify/themekit/issues/188
- fixes https://github.com/Shopify/themekit/issues/97
- fixes https://github.com/Shopify/themekit/issues/79
- fixes https://github.com/Shopify/themekit/issues/209

### Throttling

Before this change, many of the commands didn't really adhere to throttling properly. This caused a lot of errors, especially the `too many requests` errors. These should be cleared up by this PR as well. 

- fixes https://github.com/Shopify/themekit/issues/136

@chrisbutcher @ilikeorangutans @jessicaxiejw

### CODE REVIEW

File: commands/replace.go [hunk 1]
```diff
@@ -9,44 +9,27 @@ import (
...
+		}
+		for _, asset := range args.ThemeClient.LocalAssets(root) {
+			assetsActions[asset.Key] = kit.NewUploadEvent(asset)
+		}
+	} else {
+		for _, filename := range args.Filenames {
 			asset, err := theme.LoadAsset(root, filename)
-			if err == nil {
-				events <- kit.NewUploadEvent(asset)
+			if err == nil && asset.IsValid() {
```
Comment (chrisbutcher) [inline commands/replace.go]: @tanema Should we handle the `err != nil` case here as well?

File: commands/upload.go [hunk 2]
```diff
@@ -9,25 +9,22 @@ import (
...
+				foreman.JobQueue <- kit.NewUploadEvent(asset)
+			}
 		}
 	} else {
-		for _, filename := range filenames {
+		for _, filename := range args.Filenames {
 			asset, err := theme.LoadAsset(root, filename)
-			if err == nil {
-				events <- kit.NewUploadEvent(asset)
+			if err == nil && asset.IsValid() {
```
Comment (chrisbutcher) [inline commands/upload.go]: Same question. Should we handle non-nil err?

File: kit/file_watcher.go [hunk 3]
```diff
@@ -15,20 +13,31 @@ import (
...
-
-var assetLocations = []string{"templates/customers/", "assets/", "config/", "layout/", "snippets/", "templates/", "locales/", "sections/"}
-
-// FsAssetEvent ... TODO
-type FsAssetEvent struct {
-	asset     theme.Asset
-	eventType EventType
-}
+const (
+	debouceTimeout = 50 * time.Millisecond
```
Comment (chrisbutcher) [inline kit/file_watcher.go]: typo 🤓

**Comment by chrisbutcher** [comment]
A few minor comments, otherwise LGTM!

Can we get @m-ux @cshold @stevebosworth @macdonaldr93 to try a build off master once this merges this week?

File: kit/file_watcher.go (hunk 3 above)
Comment (tanema) [inline kit/file_watcher.go]: Fixed.

File: commands/replace.go (hunk 1 above)
Comment (tanema) [inline commands/replace.go]: Fixed in both.

**Comment by chrisbutcher** [comment]
🚢

**Pull request merged**